frame/contracts/src/tests.rs

```diff
@@ -1919,3 +1918,176 @@ fn instantiate_return_code() {
 
 	});
 }
+
+#[test]
+fn lazy_removal_works() {
+	let (code, hash) = compile_module::<Test>("self_destruct").unwrap();
+	ExtBuilder::default().existential_deposit(50).build().execute_with(|| {
+		let subsistence = ConfigCache::<Test>::subsistence_threshold_uncached();
+		let _ = Balances::deposit_creating(&ALICE, 10 * subsistence);
+		assert_ok!(Contracts::put_code(Origin::signed(ALICE), code));
+
+		assert_ok!(
+			Contracts::instantiate(
+				Origin::signed(ALICE),
+				subsistence,
+				GAS_LIMIT,
+				hash.into(),
+				vec![],
+				vec![],
+			),
+		);
+
+		let addr = Contracts::contract_address(&ALICE, &hash, &[]);
+		let info = <ContractInfoOf::<Test>>::get(&addr).unwrap().get_alive().unwrap();
+		let trie = &info.child_trie_info();
+
+		// Put value into the contracts child trie
+		child::put(trie, &[99], &42);
+
+		// Terminate the contract
+		assert_ok!(Contracts::call(
+			Origin::signed(ALICE),
+			addr.clone(),
+			0,
+			GAS_LIMIT,
+			vec![],
+		));
+
+		// Contract info should be gone
+		assert!(!<ContractInfoOf::<Test>>::contains_key(&addr));
+
+		// But value should be still there as the lazy removal did not run, yet.
+		assert_matches!(child::get(trie, &[99]), Some(42));
+
+		// Run the lazy removal
+		Contracts::on_initialize(0);
+
+		// Value should be gone now
+		assert_matches!(child::get::<i32>(trie, &[99]), None);
+	});
+}
+
+#[test]
+fn lazy_removal_does_not_use_all_weight() {
```

Comment:
Need test for "lazy_removal_can_partially_remove", or something like that

Reply:
Yes. However, it is not possible to write such a test right now. We need a way to commit the overlay to the backend during testing. Just like we do during benchmarking. Not sure how to do that, though.

Reply:
`ExtBuilder` uses `TestExternalities`, there is `TestExternalities::commit_all` which returns a new backend with all key/values. Maybe we should just add a method:
```diff
diff --git a/primitives/state-machine/src/testing.rs b/primitives/state-machine/src/testing.rs
index 4dcd30828..4a45b9877 100644
--- a/primitives/state-machine/src/testing.rs
+++ b/primitives/state-machine/src/testing.rs
@@ -153,6 +153,11 @@ impl<H: Hasher, N: ChangesTrieBlockNumber> TestExternalities<H, N>
                &mut self.changes_trie_storage
        }
 
+       /// Reset backend to the given one.
+       pub fn set_backend(&mut self, backend: InMemoryBackend<H>) {
+               self.backend = backend;
+       }
+
        /// Return a new backend with all pending value.
```

(Or maybe add a function similar to `TestExternalities::insert` but for child tries. But I'm not sure if there is some stuff to be careful like updating the childtrie root hash in the main trie)

Reply:
I think this PR can help you https://github.com/paritytech/substrate/pull/7808

Reply:
Added the bespoken test.